malfeasance/publisher.go

```diff
@@ -0,0 +1,74 @@
+package malfeasance
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/malfeasance/wire"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+)
+
+type Publisher struct {
+	logger    *zap.Logger
+	cdb       *datastore.CachedDB
+	tortoise  tortoise
+	sync      syncer
+	publisher pubsub.Publisher
+}
+
+func NewPublisher(
+	logger *zap.Logger,
+	cdb *datastore.CachedDB,
+	sync syncer,
+	tortoise tortoise,
+	publisher pubsub.Publisher,
+) *Publisher {
+	return &Publisher{
+		logger:    logger,
+		cdb:       cdb,
+		tortoise:  tortoise,
+		sync:      sync,
+		publisher: publisher,
+	}
+}
+
+// Publishes a malfeasance proof to the network.
+func (p *Publisher) PublishProof(ctx context.Context, smesherID types.NodeID, proof *wire.MalfeasanceProof) error {
+	malicious, err := identities.IsMalicious(p.cdb, smesherID)
+	if err != nil {
+		return fmt.Errorf("check if smesher is malicious: %w", err)
+	}
+	if malicious {
+		p.logger.Debug("smesher is already marked as malicious", zap.String("smesher_id", smesherID.ShortString()))
+		return nil
+	}
+
+	if err := identities.SetMalicious(p.cdb, smesherID, codec.MustEncode(proof), time.Now()); err != nil {
+		return fmt.Errorf("adding malfeasance proof: %w", err)
+	}
+
+	p.cdb.CacheMalfeasanceProof(smesherID, codec.MustEncode(proof))
+	p.tortoise.OnMalfeasance(smesherID)
+
+	// Only gossip the proof if we are synced (to not spam the network with proofs others probably already have).
+	if !p.sync.ListenToATXGossip() {
+		p.logger.Debug("not synced, not broadcasting malfeasance proof",
+			zap.String("smesher_id", smesherID.ShortString()),
+		)
+		return nil
+	}
```

Comment:
Perhaps the proof could be put into some queue for publishing later when the node becomes synced.

Reply:
I am not sure that would be a good idea:

- we are not syncing malfeasance proofs before we are ATX synced because we don't consider proofs for identities we have never seen as valid (otherwise it would be possible to spam). Right now we are dropping a peer sending us a proof of an identity that doesn't exist (according to our view on the network) to avoid a possible spam vector.
- we would need to start ignoring proofs of identities we don't know. But then we would keep receiving the same proof every 30 minutes (malfeasance sync interval) that we do not store until we receive the first ATX of that identity. This becomes another spam vector.
- so we have to not sync malfeasance information until we are in ATX sync and every malfeasance we detect on the way we only persist and do not gossip because everyone who is ATX synced should already know about it.
- when we are ATX synced we should gossip a malfeasance proof so that others who consider themselves ATX synced but haven't yet received all the information needed to detect the malfeasance themselves are informed.

Reply:
I agree with @fasmat here, but there's a bit of issue with naming of `PublishProof` function.
In case if `ListenToATXGossip()` returns false, what this func actually does is marking the identity as malicious locally, and not publish anything. It may also do nothing at all if the identity is already marked as malicious in the database.
I'm unsure about proper function name though. `StorePublishProof`, `EnsureMalicious`... may need some thinking